README.md
[![ko-fi](https://img.shields.io/badge/-buy%20me%20a%20coffee-%23FF5E5B?style=plastic)](https://ko-fi.com/slate) [![patreon](https://img.shields.io/badge/-support%20me%20on%20patreon-%235C5C5C?style=plastic)](https://patreon.com/slatesfoundrystuff) ![GitHub release (by tag)](https://img.shields.io/github/downloads/zarmstrong/hlo-importer/hlo-importer-0.10.1/total?style=plastic) ![GitHub all releases](https://img.shields.io/github/downloads/zarmstrong/hlo-importer/total?style=plastic) ![GitHub](https://img.shields.io/github/license/zarmstrong/hlo-importer?style=plastic)

# Herolab Online PF2e Import Tool

A module for Foundry VTT that converts and imports Herolab Online 2e characters.

Please be aware that this module attempts to match Herolab Online items to Foundry items, and will not perform perfectly in all cases due to database differences.

## Navigation
1. [Important Note](#important-note)
2. [Guide for Importing from Herolab Online](#guide-for-importing-from-herolab-online)
3. [Getting your User Token from HLO](#getting-your-user-token-from-hlo)
4. [Getting the Element Token from HLO](#getting-the-element-token-from-hlo)
5. [Special Thanks](#special-thanks)
6. [License](#license)

## Important Note

This importer relies on my other tool, [Herolab to Foundry VTT PF2](https://github.com/zarmstrong/hlo-to-fvtt-pf2-public) to do the actual conversion process. If you have problems with the conversion, please be sure to [open a bug report on that project](https://github.com/zarmstrong/hlo-to-fvtt-pf2-public/issues).

## Guide for Importing from Herolab Online

1) Install this module or get your GM to do it in Foundry VTT.
2) Get the Element Token for the character you wish to import 
3) In *Foundry VTT* click Import from HLO, found right at the top of your character sheet next to the close button.
4) Enter the Element Token you got from Herolab Online and click import.
5) Wait about 30-45 seconds.
6) Verify the character name, class and level match your expectations. Take note of any messages from the conversion process, as you may need to make manual changes. When you're ready, click Proceed.
***
## Getting your User Token from HLO

(NOTE: If your character(s) are in a campaign, the steps are slightly different than below. In the campaign, open the character and click on the gear in the top right corner. Then follow the steps below starting at step 2)

Step 1: Click the Kebab menu on the character you wish to export<br>
![Click the Kebab menu on the character you wish to export](get-element-id-step1.webp)

Step 2: Click Export/Integrate<br>
![Click Export/Integrate](get-element-id-step2.webp)

Step 3: Click Get User Token<br>
![Click Get User Token](get-user-token-step3.webp)

Step 4: Click Copy to Clipboard<br>
![Click Copy to Clipboard](get-user-token-step4.webp)
***
## Getting the Element Token from HLO

(NOTE: If your character(s) are in a campaign, the steps are slightly different than below. In the campaign, open the character and click on the gear in the top right corner. Then follow the steps below starting at step 2)

Step 1: Click the Kebab menu on the character you wish to export<br>
![Click the Kebab menu on the character you wish to export](get-element-id-step1.webp)

Step 2: Click Export/Integrate<br>
![Click Export/Integrate](get-element-id-step2.webp)

Step 3: Click Get Element Token<br>
![Click Get Element Token](get-element-id-step3.webp)

Step 4: Click Copy to Clipboard<br>
![Click Copy to Clipboard](get-element-id-step4.webp)

*** 
#### Special Thanks

Special thanks to @Doctor-Unspeakable for his [Pathbuilder module](https://github.com/Doctor-Unspeakable/foundry-pathbuilder2e-import) for being the basis of this Foundry module.

### License
This Foundry VTT module is licensed under the [MIT License](https://github.com/zarmstrong/hlo-importer/blob/main/LICENSE).
This work is licensed under [Foundry Virtual Tabletop EULA - Limited License Agreement for module development](https://foundryvtt.com/article/license/).


CHANGELOG.md
## 0.10.1 (September 15, 2022)

* minor change to compatability info

## 0.10.0 (September 15, 2022)

* overhauled for Foundry v10 and pf2e 4.x

## 0.8.5 (May 25, 2022)

* no code change, just updated the module to show compatability with v9

## 0.8.4 (Mar 26, 2022)

* prevent module from showing up in non-pf2e systems

## 0.8.3 (Jan 1, 2022)

* fix permissions on import

## 0.8.2 (October 10, 2021)

* updated codebase for PF2e 2.15. 

## 0.8.1 (August 21, 2021)

* update for 0.8.9 compatability
* introduced compatability with http instead of https (should work with both)

## 0.8.0 (August 11, 2021)

* updated codebase for PF2e 2.11. All users are required to upgrade as there are breaking changes

## 0.7.5 (July 20, 2021)

* fixed bug that would export characters to herovault that weren't supposed to be

## 0.7.4 (July 3, 2021)

* removed extraneous code

## 0.7.3 (July 3, 2021)

* fixed a bug that would export the PC to herovault even when the export to herovault box was unchecked.
* fixed a bug triggered by importing PC over an existing PC

## 0.7.2 (June 30, 2021)

* version bump to fix bad commit

## 0.7.1 (June 30, 2021)

* fixed the connector I rewrote. forgot to include the change 

## 0.7.0 (June 30, 2021)

* rewrote the connectors between hlo importer and herovault to use native api calls instead of janky crap

## 0.6.2 (June 30, 2021)

* revert previous change

## 0.6.1 (June 30, 2021)

* change relative paths to absolute for imports

## 0.6.0 (June 28, 2021)

* This module now integrates with HeroVau.lt and I highly recommend you install it, too!
* upped compatible version to 0.8.8
* prevent herolab button from showing if herovault is present and active

## 0.5.3 (June 23, 2021)

* fixed HLO importer showing up on non-character sheets
* cleaned up some code

## 0.5.2 (June 22, 2021)

* support for foundry 0.8.7

## 0.5.1 (June 18, 2021)

* minify the javascript
* added function in build code to ensure debug is never deployed

## 0.5.0 (June 3, 2021)

* Support for foundry 0.8.6+ only

## 0.4.1 (May 15, 2021)

* Fixed incompatability with the Foundry native app (thanks @Drental)

## 0.4.0 (May 12, 2021)

* Feats that have effects attached to them are now applied to the PC upon import.
* New tokens will have the appropriate actor ID

## 0.3.4 (May 5, 2021)

* Fix minor error in console output
* Added a setting to enable/disable debug output so source code editing wouldn't be required.

## 0.3.3 (Apr 30, 2021)

* Method implemented in 0.3.2 to prevent module from being loaded doesn't quite work, even though it should according to the documentation. Added a check in the code to prevent it from working outside of pf2e.

## 0.3.2 (Apr 30, 2021)

* prevent module from being loaded in other game systems (thanks wildj79)

## 0.3.1 (Apr 28, 2021)

* fix bug that was preventing error messages from showing

## 0.3.0 (Apr 28, 2021)

* add hlo importer and gamesystem version to js
* add color to console logging
* add hlo importer and gamesystem version to requests to check for compatability breaking features

## 0.2.5 (Mar 12, 2021)

* update to support PF2e 1.11.13

## 0.2.4 (Mar 01, 2021)

* add support for bug reporter module

## 0.2.3 (Feb 14, 2021)

* fix cancel button not working

## 0.2.2 (Jan 4, 2021)

* disable debug mode

## 0.2.1 (Jan 3, 2021)

* Made a proper release

## 0.2.0 (Jan 3, 2021)

* Added support for GMs to use their own User Token instead of my User Token (default). This value can be set through the module settings page. It is currently optional. I expect it will be mandatory at some time in the future.
* Added rudamentary error handling

## 0.1.2 (Jan 3, 2021)

* Change function names; I think they were possibly conflicting with Import from Pathbuilder 2e
* add debug flag to enable debug output to console

## 0.1.1 (Jan 2, 2021)

* Change Import icon to flask to imitate the flask that HLO uses for it's favicon. 
* Changed the menu title to HLO (from Import from HLO) to save some space. Title bars get pretty crowded if you have a lot of other modules that use that space.
* Prevent Import button from showing up on anything other than character sheets

## 0.1.0 (Jan 2, 2021)

* Initial Release



hlo-importer.js
let hlodebug = false;
const hloiVer="0.10.1";
let herolabURL="https://www.pf2player.com";

const color1='color: #7bf542';  //bright green
const color2='color: #d8eb34'; //yellow green
const color3='color: #ffffff'; //white
const color4='color: #cccccc'; //gray
const color5='color: #ff0000'; //red
var hlo,userToken;
var hloButton=true;
if (location.protocol !== "https:") {
  herolabURL = "http://www.pf2player.com";
}

Hooks.on('ready', async function() {
  if (game.system.id!="pf2e") {
    console.log("%cHLO Importer | %cWrong game system. %cNot enabling.",color1,color5,color4);
  } else {
    console.log("%cHLO Importer | %cinitializing",color1,color4);
      game.settings.register('hlo-importer', 'userToken', {
          name : "User Token (optional)",
          hint : "Please enter your personal user token. A user token allows external tools (like this one) to access the HLO server and perform export operations.",
          scope : 'world',
          config : true,
          type : String,
          default : '',
          onChange: value => (userToken=game.settings.get('hlo-importer', 'userToken'))
      });
      game.settings.register('hlo-importer', 'debugEnabled', {
          name : "Enable debug mode",
          hint : "Debug output will be written to the js console.",
          scope : 'world',
          config : true,
          type: Boolean,
          default: false,
          onChange: value => (hlodebug=game.settings.get('hlo-importer', 'debugEnabled'))
      });
  }
  hlo = new HeroLabImporter(hlodebug);
  hlodebug=game.settings.get('hlo-importer', 'debugEnabled');
  userToken=game.settings.get('hlo-importer', 'userToken')
});

Hooks.on('herovaultfoundryReady', (api) => {
  if (hlodebug)
    console.log("Disabling HLO button since herovault is loaded");
  hloButton=false;
});

Hooks.on('renderActorSheet', function(obj, html){
  hlodebug = game.settings.get('hlo-importer', 'debugEnabled');
  if (game.system.id!="pf2e") {
    console.log("%cHLO Importer | %cWrong game system. %cNot adding HLO button to actor sheet.",color1,color5,color4);
  } else {
    // Only inject the link if the actor is of type "character" and the user has permission to update it
      const actor = obj.actor;
      if (hlodebug) {
        console.log("%cHLO Importer | %cPF2e System Version: hlo-importer actor type: " + actor.type,color1,color4);
        console.log("%cHLO Importer | %cCan user modify: " + actor.canUserModify(game.user, "update"),color1,color4);
      }

      if (!(actor.type === "character")){ return;}
      if (actor.canUserModify(game.user, "update")==false){ return;}
      
      if (hloButton) {
        let element = html.find(".window-header .window-title");
        if (element.length != 1) {return;}
        
        let button = $(`<a class="popout" style><i class="fas fa-flask"></i>HLO</a>`);
        userToken = game.settings.get('hlo-importer', 'userToken');
        if (hlodebug) {
          console.log("%cHLO Importer | %chlo-importer token: "+ userToken,color1,color4);
        }
        button.on('click', () => hlo.beginHLOImport(obj.object,userToken));
        element.after(button);
      }
    }
  }
);

async function doHVExport(hero,act) {
  game.modules.get('herovaultfoundry')?.api?.exportToHVFromPBHLO(hero,act);
  // if (game.modules.get('herovaultfoundry')?.active){
  //   let {exportToHVFromPBHLO} = await import('../herovaultfoundry/herovault-min.js');
  //   if (typeof exportToHVFromPBHLO !== "undefined") {
  //     exportToHVFromPBHLO(hero,act);
  //   }
  // }  
  // return;
}

export function hloShim(targetActor) {
   let hlo = new HeroLabImporter;
    hlo.heroVaultPrompt=true;
    userToken = game.settings.get('hlo-importer', 'userToken');
    let x = hlo.beginHLOImport(targetActor,userToken);
}

export function hloActive() {
  return true;
}

export class HeroLabImporter {
  constructor(hlodebug) {
    this.color1='color: #7bf542';  //bright green
    this.color2='color: #d8eb34'; //yellow green
    this.color3='color: #ffffff'; //white
    this.color4='color: #cccccc'; //gray
    this.color5='color: #ff0000';
    this.hlodebug = hlodebug;
    this.heroVaultExport=false;
    this.heroVaultPrompt=false;
  }

  beginHLOImport(targetActor,userToken){
    let applyChanges=false;
    let hvCheckbox="";
    if (this.heroVaultPrompt) {
      hvCheckbox='<br><input type="checkbox" id="checkBoxHVExport" name="checkBoxHVExport" ><label for="checkBoxHVExport"> Export this PC to my HeroVau.lt</label><br>';
    }
    new Dialog({
      title: `Herolab Online Import`,
      content: `
        <div>
          <p>Step 1: Get the character token by clicking on the kebab menu (<strong>⋮</strong>) on any character on your account. Scroll down to "Element Token" and click the <strong>Get Element Token</strong> button. Click the <strong>Copy to Clipboard</strong> button.</p>
          <p>Step 2: Paste the Element Token from the Herolab Online export dialog below</p>
          <br>
          <p>Please note - items which cannot be matched to the Foundry database will not be imported!<p>
        <div>
        <hr/>
        <div id="divCode">
          Enter the element token of the character you wish to import<br>
          <div id="divOuter">
            <div id="divInner">
              <input id="textBoxElementID" type="text" maxlength="14" />
            </div>
          </div>
        </div>
        ${hvCheckbox}
        <br><br><strong>Once you click Import, please be patient as the process might take up to 45 seconds to complete.</strong><br><br>
        <style>
        
          #textBoxElementID {
              border: 0px;
              padding-left: 5px;
              letter-spacing: 2px;
              width: 330px;
              min-width: 330px;
            }
            
            #divInner{
              left: 0;
              position: sticky;
            }
            
            #divOuter{
              width: 285px; 
              overflow: hidden;
            }
    
            #divCode{  
              border: 1px solid black;
              width: 300px;
              margin: 0 auto;
              padding: 5px;
            }
    
        </style>
        `,
      buttons: {
        yes: {
          icon: "<i class='fas fa-check'></i>",
          label: `Import`,
          callback: () => applyChanges = true
        },
        no: {
          icon: "<i class='fas fa-times'></i>",
          label: `Cancel`
        },
      },
      default: "yes",
      close: html => {
        if (applyChanges) {
          let HLOElementID= html.find('[id="textBoxElementID"]')[0].value;

          if (html.find('[id="checkBoxHVExport"]').length) {
            if (hlodebug)
              console.log("%cHLO Importer | %cevalue of hloexport check: "+html.find('[id="checkBoxHVExport"]')[0].value,color1,color4);
            if (html.find('[id="checkBoxHVExport"]')[0].checked)
              this.heroVaultExport=true;
            else
              this.heroVaultExport=false;
          }

          this.convertHLOCharacter(targetActor, HLOElementID,userToken);
        }
      }
    }).render(true);

  }

  convertHLOCharacter(targetActor, HLOElementID, userToken){
      const pf2eVersion=game.system.version;
      let error=false;
      var self=this;
      var xmlhttp = new XMLHttpRequest();
      xmlhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          let responseJSON = JSON.parse(this.responseText);
          if (hlodebug){ 
            console.log("%cHLO Importer | %c"+responseJSON,color1,color4);
          }
          if (responseJSON.hasOwnProperty("error")) {
            if (hlodebug) {
               console.log("%cHLO Importer | %cerror found in response",color1,color4);
             }
            error=true;
          } else {
            if (hlodebug) {
              console.log("%cHLO Importer | %c"+Object.keys(responseJSON.characterData).length,color1,color4);
            }
          }

          if (error){
            new Dialog({
              title: `Herolab Online Import`,
              content: `
                   <div>
                      <h3>Error</h3>
                      <p>${responseJSON.error}<p>
                   </div><br>`,
              buttons: {
                yes: {
                  icon: "<i class='fas fa-check'></i>",
                  label: `Ok`              }
              },
              default: "yes"
            }).render(true);
          } else {
            if (Object.keys(responseJSON.characterData).length>1){
              if (hlodebug) {
                console.log("%cHLO Importer | %cCalling checkHLOCharacterIsCorrect",color1,color4);
              }
              self.checkHLOCharacterIsCorrect(targetActor, responseJSON);
            } else {
              ui.notifications.warn("Unable to convert. Please file a bug with the Conversion ID: " + responseJSON.ConversionID); 
         }
        }
          
        // console.log("%cHLO Importer | %creadyState: "+this.readyState,color1,color4)
        }
      };
      if (hlodebug) {
        console.log("%cHLO Importer | %cusertoken: " + userToken,color1,color4);
        console.log("%cHLO Importer | %cPF2e System Version: " + pf2eVersion,color1,color4);
      }
      if (userToken == "") {
        xmlhttp.open("GET", herolabURL + "/foundrymodule.php?elementID="+encodeURIComponent(HLOElementID)+"&pf2e="+pf2eVersion+"&hloi="+hloiVer, true);
      }
      else {
        xmlhttp.open("GET", herolabURL + "/foundrymodule.php?elementID="+encodeURIComponent(HLOElementID)+"&pf2e="+pf2eVersion+"&hloi="+hloiVer+"&userToken="+encodeURIComponent(userToken), true);
      }
      xmlhttp.send();
  }

  checkHLOCharacterIsCorrect(targetActor,responseJSON){
    if (hlodebug ){
      console.log("%cHLO Importer | %cin checkHLOCharacterIsCorrect",color1,color4);
      console.log("%cHLO Importer | %c",color1,color4);
      console.log(responseJSON);
    }
    let correctCharacter = false;
    let charImport = responseJSON.characterData;
    const conversionData=responseJSON.conversionData;
    new Dialog({
      title: charImport.name,
      content: `
        <div><h2>Conversion Log:<br>`+responseJSON.conversionData+`</div><br><div><strong>Continue importing `+charImport.name+`, level `+charImport.system.details.level.value+` `+charImport.flags.herolabimporter.class+`?</strong></div><br><br>
        `,
      buttons: {
        yes: {
          icon: "<i class='fas fa-check'></i>",
          label: `Proceed`,
          callback: () => correctCharacter = true
        },
        no: {
          icon: "<i class='fas fa-times'></i>",
          label: `Cancel`
        },
      },
      default: "yes",
      close: html => {
        if (correctCharacter) {
          this.importHLOCharacter(targetActor, charImport);
        }
      }
    }).render(true);
  }

  async importHLOCharacter(targetActor, charImport){
    // let importPCID=new RegExp(charImport._id, "g");
    // let targetPCID=targetActor.data._id;
    let charDataStr=JSON.stringify(charImport);
    // charDataStr=charDataStr.replace(importPCID,targetPCID);
    charImport=JSON.parse(charDataStr);
    if (hlodebug) {
      console.log("%cHLO Importer | %c Importing "+charImport.name,color1,color4);
      console.log("%cHLO Importer | %c HV export: "+this.heroVaultExport,color1,color4);
    }
    let oldPermissions =targetActor.permission;
    charImport.permission = oldPermissions;
    if (hlodebug)
      console.log("%cHLO Importer | %c deleting embedded documents",color1,color4)

    await targetActor.deleteEmbeddedDocuments('Item', [],{deleteAll: true});
    if (hlodebug)
      console.log("%cHLO Importer | %c running importFromJSON",color1,color4)
    await targetActor.importFromJSON(JSON.stringify(charImport));
    if (hlodebug)
      console.log("%cHLO Importer | %c importFromJSON complete",color1,color4)
    targetActor.update({
      "flags.exportSource.world": game.world.id,
      "flags.exportSource.system": game.system.id,
      "flags.exportSource.systemVersion": game.system.version,
      "flags.exportSource.coreVersion": game.version,
      "flags.herolabimporter.version.value": hloiVer,
    });
    if (this.heroVaultExport) {
      if (hlodebug)
        console.log("%cHLO Importer | %cperforming a HeroVau.lt Export",color1,color4);
      doHVExport(JSON.stringify(charImport),targetActor)
    }
  }
}

Hooks.on('init', () => {
    game.modules.get('hlo-importer').api = {
    hloShim: hloShim,
    hloActive: hloActive
  };
  Hooks.callAll('hloimporterReady', game.modules.get('hlo-importer').api);
});